File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return dfs(root, Integer.MIN_VALUE);
    }

    public int dfs(TreeNode root, int max) {
        if(root == null) return 0;
        
        int good = root.val >= max ? 1 : 0;
        int currMax = Math.max(max, root.val);
        return good += dfs(root.left, currMax) + dfs(root.right, currMax); 
    }

    // public int goodNodes(TreeNode root) {
    //     int[] res = {0};
    //     dfs(root, Integer.MIN_VALUE, res);
    //     return res[0];
    // }

    // public void dfs(TreeNode root, int max, int[] res) {
    //     if(root == null) return;
    //     if(max <= root.val) res[0]++;
    //     max = Math.max(max, root.val);
    //     dfs(root.left, max, res);
    //     dfs(root.right, max, res);
    //     // max(max, root.val)
    // }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [-2**31]
        self.dfs(root, res)
        return res[0]
    
    def dfs(self, root: TreeNode, res: list[int]) -> int:
        if not root:
            return 0

        left = max(0, self.dfs(root.left, res))
        right = max(0, self.dfs(root.right, res))

        maxVal = left + right + root.val 
        res[0] = max(res[0], maxVal)
        best = max(left, right)
        best = root.val + best if root.val + best > root.val else root.val

        return best